File: sdk/dart/dovahlink_client/lib/src/protocol/session_invalidated_payload.dart
import 'package:json_annotation/json_annotation.dart';

import '../shared/enums.dart';
import 'json_map.dart';
import 'protocol_format_exception.dart';

part 'session_invalidated_payload.g.dart';

/// Incoming `session_invalidated` payload (`protocol/schema/README.md`'s `session_invalidated`).
/// Decode-only: the client never sends `session_invalidated`.
@JsonSerializable(checked: true, createToJson: false)
class SessionInvalidatedPayload {
  /// Creates a session-invalidated payload.
  const SessionInvalidatedPayload({required this.reason});

  /// Decodes and validates one `session_invalidated` payload.
  factory SessionInvalidatedPayload.fromJson(JsonMap json) {
    try {
      return _$SessionInvalidatedPayloadFromJson(json);
    } on Object catch (error) {
      throw ProtocolFormatException(
        'Invalid session_invalidated payload: $error',
      );
    }
  }

  /// The wire vocabulary of `session_invalidated.reason`.
  @JsonKey(required: true)
  final AdministrativeInvalidationReason reason;
}

File: integration/DovahLinkValidationClient/SessionInvalidatedPayload.cs
using System.Text.Json.Nodes;

namespace DovahLinkValidationClient;

/// <summary>
/// Incoming <c>session_invalidated</c> payload (<c>protocol/schema/README.md</c>'s
/// <c>session_invalidated</c>). Decode-only: this client never sends <c>session_invalidated</c>.
/// An unsolicited, best-effort terminal event for an administratively invalidated session.
/// </summary>
/// <param name="Reason">The wire vocabulary of <c>session_invalidated.reason</c>: <c>"revoked"</c>,
/// <c>"blocked"</c>, <c>"trust_reset"</c>, or <c>"factory_reset"</c>.</param>
public sealed record SessionInvalidatedPayload(string Reason)
{
    /// <summary>
    /// Decodes and validates one <c>session_invalidated</c> payload.
    /// </summary>
    /// <param name="payload">The envelope's decoded payload object.</param>
    /// <returns>The decoded session-invalidated payload.</returns>
    /// <exception cref="FormatException">Thrown when reason is missing, has the wrong JSON type, or
    /// is not a registered invalidation reason.</exception>
    public static SessionInvalidatedPayload Decode(JsonObject payload)
    {
        try
        {
            string reason = payload["reason"]?.GetValue<string>() ?? throw new FormatException("Missing reason.");
            if (reason is not ("revoked" or "blocked" or "trust_reset" or "factory_reset"))
            {
                throw new FormatException($"Unknown session invalidation reason: {reason}.");
            }
            return new SessionInvalidatedPayload(reason);
        }
        catch (InvalidOperationException ex)
        {
            throw new FormatException($"Malformed session_invalidated payload: {ex.Message}", ex);
        }
    }
}
